File: 84.largest-rectangle-in-histogram.go
/*
 * @lc app=leetcode id=84 lang=golang
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
func largestRectangleArea(heights []int) int {
	maxArea, stack, height := 0, []int{}, 0

	for i := 0; i <= len(heights); i++ {
		if i == len(heights) {
			height = 0
		} else {
			height = heights[i]
		}

		if len(stack) == 0 || height >= heights[stack[len(stack)-1]] {
			stack = append(stack, i)
		} else {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			length := 0
			if len(stack) == 0 {
				length = i
			} else {
				length = i - 1 - stack[len(stack)-1]
			}

			maxArea = max(maxArea, length*heights[top])
			i--
		}
	}

	return maxArea
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

File: 84.largest-rectangle-in-histogram.cpp
/*
 * @lc app=leetcode id=84 lang=cpp
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
class Solution {
 public:
  int largestRectangleArea(vector<int>& heights) {
    vector<int> H(heights);
    H.insert(H.begin(), 0);
    H.push_back(0);

    stack<int> st;
    st.push(0);

    int ans = 0;
    for(int i = 1; i < H.size(); ++i) {
      while(H[i] < H[st.top()]) {
        int mid = st.top();
        st.pop();

        int w = i - st.top() - 1;
        int h = H[mid];
        ans = max(ans, w * h);
      }

      st.push(i);
    }

    return ans;
  }
};
// @lc code=end

